### PR TITLE
add pair_wise_loss and pit_loss_from_pair_wise

### PR DESCRIPTION
**Pull request by boeddeker**

### CODE REVIEW

**Comment by boeddeker** [comment]
@TCord Can you test, if this code works in your case?

File: padertorch/ops/losses/source_separation.py [hunk 1]
```diff
@@ -105,18 +105,194 @@ def pit_loss(
...
+        del estimate_shape[1]
+        assert estimate_shape == list(target.shape), (
+            f'{estimate.shape} (N, K, ...) does not match {target.shape} (N, ...)'
+        )
+    else:
+        assert estimate.size() == target.size(), (
+            f'{estimate.size()} != {target.size()}'
+        )
+
+    pair_wise_loss = []
```
Comment (thequilo) [inline padertorch/ops/losses/source_separation.py:214]: This is bad naming because the function has the same name. You could just call it `loss` or `pair_wise_losses` because it collects loss values of all pairs

File: padertorch/ops/losses/source_separation.py [hunk 2]
```diff
@@ -105,18 +105,194 @@ def pit_loss(
...
+            'nc...,n...k->n...ck',
+            -torch.nn.LogSoftmax(dim=1)(estimate),
+            torch.nn.functional.one_hot(target, num_classes=sources).to(estimate.dtype)
+        ), 'n ... c k -> c k', reduction='mean')
+
+    return pair_wise_loss
+
+
+def pit_loss_from_pair_wise(
+        pair_wise_loss,
```
Comment (thequilo) [inline padertorch/ops/losses/source_separation.py:243]: Again, there already is a function called `pair_wise_loss`

File: padertorch/ops/losses/source_separation.py [hunk 3]
```diff
@@ -92,7 +92,7 @@ def pit_loss(
         >>> pit_loss(estimate, target, axis=-3)
         tensor(1.)
     """
-    axis = axis % estimate.ndimension()
+    # axis = axis % estimate.ndimension()
```
[the remark this answers is not in the document]
Comment (boeddeker) [reply padertorch/ops/losses/source_separation.py]: Done

File: padertorch/ops/losses/source_separation.py [hunk 4]
```diff
@@ -105,18 +105,194 @@ def pit_loss(
...
+            f'{estimate.shape} (N, K, ...) does not match {target.shape} (N, ...)'
+        )
+    else:
+        assert estimate.size() == target.size(), (
+            f'{estimate.size()} != {target.size()}'
+        )
+
+    pair_wise_loss = []
+
+    if estimate.ndim == target.ndim:
```
[the remark this answers is not in the document]
Comment (boeddeker) [reply padertorch/ops/losses/source_separation.py]: Done

File: padertorch/ops/losses/source_separation.py (hunk 1 above)
Comment (boeddeker) [reply padertorch/ops/losses/source_separation.py]: I removed this variable in this function (Was from an older version of the implementation)

File: padertorch/ops/losses/source_separation.py (hunk 2 above)
Comment (boeddeker) [reply padertorch/ops/losses/source_separation.py]: I changed it to `pair_wise_loss_matrix`.
I wanted to keep the name verbose (i.e. obvious connection to `pair_wise_loss` function).

**Comment by boeddeker** [comment]
In general, I am open for better names for the functions `pair_wise_loss` and `pit_loss_from_pair_wise`.

**Comment by TCord** [comment]
Maybe you can change the `pair_wise_loss ` function into something like `compute_pairwise_losses`
Aside from that, it seems to work for my application with the `BCEwithlogits` loss function.

File: padertorch/ops/losses/source_separation.py (hunk 2 above)
Comment (TCord) [reply padertorch/ops/losses/source_separation.py]: Maybe you can also change `pit_loss_from_pair_wise` to `pit_loss_from_loss_matrix`. I think this way the usage of this function is clearer.

**Review by TCord — APPROVED** [review]
